### PR TITLE
Add slu snips data receipe

### PR DESCRIPTION
**Pull request by yuekaizhang**
Currently, for text style: (follwing this paper https://arxiv.org/pdf/1809.09190.pdf)
`wav_name INTENT ENTITY SLOT_NAME slot1_value ENTITY SLOT_NAME slot2_value`
`01356b70-791d-42c4-b14b-56fefa67358f-1622.wav SETLIGHTCOLOR HOUSE_ROOM_UNIQUE ROOM bedroom COLOR COLOR red`

The next step would be training. Any suggestions about it? 
The training set is very small, do we need a pretrained ASR? 
Also, should we only focus on the intent classification tasks for now? (Many slot values occur infrequently in the dataset. )

Another dumb question, I am not sure I find the correct snips dataset, this recipe is for smart-light-en-closed-field subset, which has only 1660 utterances.

### CODE REVIEW

File: egs2/TEMPLATE/asr1/db.sh [hunk 1]
```diff
@@ -43,6 +43,7 @@ JVS=downloads
 KSS=
 SPGISPEECH=
 SWBD=
+SNIPS= # smart-light-en-closed-field data path
```
Comment (ftshijt) [inline egs2/TEMPLATE/asr1/db.sh:46]: Please also add it for jhu domains

File: egs2/snips/asr1/conf/tuning/train_asr_transformer_adam.yaml [hunk 2]
```diff
@@ -0,0 +1,31 @@
+# network architecture
```
Comment (ftshijt) [inline egs2/snips/asr1/conf/tuning/train_asr_transformer_adam.yaml:1]: At least for the initial PR, we could stick to one working configuration for the recipe, so please remove those not working

File: egs2/snips/asr1/run.sh [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+# -e 'error', -u 'undefined variable', -o ... 'error in pipeline', -x 'print commands',
+set -e
+set -u
+set -o pipefail
+
+train_set=train
+valid_set=dev
+test_sets="dev test"
+
+asr_config=conf/train_asr_conformer.yaml
```
Comment (ftshijt) [inline egs2/snips/asr1/run.sh:12]: This seems incompatible with the conf path?

**Review by ftshijt — COMMENTED** [review]
For a reminder, please also update `egs2/README.md` with dataset information.

**Comment by ftshijt** [comment]
> Currently, for text style: (follwing this paper https://arxiv.org/pdf/1809.09190.pdf)
> `wav_name INTENT ENTITY SLOT_NAME slot1_value ENTITY SLOT_NAME slot2_value`
> `01356b70-791d-42c4-b14b-56fefa67358f-1622.wav SETLIGHTCOLOR HOUSE_ROOM_UNIQUE ROOM bedroom COLOR COLOR red`
> 
> The next step would be training. Any suggestions about it?
> The training set is very small, do we need a pretrained ASR?
> Also, should we only focus on the intent classification tasks for now? (Many slot values occur infrequently in the dataset. )
> 
> Another dumb question, I am not sure I find the correct snips dataset, this recipe is for smart-light-en-closed-field subset, which has only 1660 utterances.

I'm not quite sure about the datasize but could you please share the link of where you get the data?

**Comment by siddhu001** [comment]
Hi @yuekaizhang can you try Spec-aug (https://github.com/espnet/espnet/blob/7bd3725ec59c1afec5346538af3c57da238d8e80/egs2/fsc/asr1/conf/tuning/train_asr_transformer_adam_small_specaug.yaml) since it really seem to improve performance on another SLU benchmark - FSC dataset (https://docs.google.com/spreadsheets/d/10pXg_Q7QIJGofquKbGUNTekouEt9gYsN9q4JGC1ycAI/edit?usp=sharing).

**Comment by siddhu001** [comment]
> Currently, for text style: (follwing this paper https://arxiv.org/pdf/1809.09190.pdf)
> `wav_name INTENT ENTITY SLOT_NAME slot1_value ENTITY SLOT_NAME slot2_value`
> `01356b70-791d-42c4-b14b-56fefa67358f-1622.wav SETLIGHTCOLOR HOUSE_ROOM_UNIQUE ROOM bedroom COLOR COLOR red`
> 
> The next step would be training. Any suggestions about it?
> The training set is very small, do we need a pretrained ASR?
> Also, should we only focus on the intent classification tasks for now? (Many slot values occur infrequently in the dataset. )
> 
> Another dumb question, I am not sure I find the correct snips dataset, this recipe is for smart-light-en-closed-field subset, which has only 1660 utterances.

Yeah, smart-light-en-closed-field subset is the correct subset for SNIPs dataset that is mostly used for benchmarking. 

We can later experiment with smart-light-en-far-field as well

**Comment by siddhu001** [comment]
> Currently, for text style: (follwing this paper https://arxiv.org/pdf/1809.09190.pdf)
> `wav_name INTENT ENTITY SLOT_NAME slot1_value ENTITY SLOT_NAME slot2_value`
> `01356b70-791d-42c4-b14b-56fefa67358f-1622.wav SETLIGHTCOLOR HOUSE_ROOM_UNIQUE ROOM bedroom COLOR COLOR red`
> 
> The next step would be training. Any suggestions about it?
> The training set is very small, do we need a pretrained ASR?
> Also, should we only focus on the intent classification tasks for now? (Many slot values occur infrequently in the dataset. )


I would suggest focusing on intent classification for now in this PR. But I think we do want to extend to slot filling as well

File: egs2/snips/asr1/run.sh [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+# speed perturbation related
+# (train_set will be "${train_set}_sp" if speed_perturb_factors is specified)
+speed_perturb_factors="0.9 1.0 1.1"
+
+./asr.sh                                               \
+    --use_lm false                                     \
+    --nlsyms_txt  "data/non_linguistic_symbols.txt"    \
+    --audio_format wav                                 \
+    --feats_type raw                                   \
+    --token_type char                                  \
```
Comment (siddhu001) [inline egs2/snips/asr1/run.sh:25]: Hi you should also try with word level decoding since that would be closer to modelling it as classification approach and also has given better performance in other SLU and Speech Dialogue Act benchmark

File: egs2/snips/asr1/local/data.sh [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+   paste -d "|" data/tmp/wavs data/tmp/semantics.sort > data/tmp/wav_sem_text
+   cut -d "|" -f 1,3 data/tmp/wav_sem_text > data/tmp/wav_sem
+   sed -r 's/\|/ /g' data/tmp/wav_sem > data/text
+   mkdir -p data/all
+   mv data/utt2spk data/text data/wav.scp data/all 
+   utils/fix_data_dir.sh data/all
+   rm -r data/tmp
+fi
+
+if [ ${stage} -le 3 ] && [ ${stop_stage} -ge 3 ]; then
```
Comment (siddhu001) [inline egs2/snips/asr1/local/data.sh:65]: Could you make sure that you are using the splits as used in "Tie your embeddings down: Cross-modal
latent spaces for end-to-end spoken language understanding," since those are the splits that have been used to benchmark SLU models.

**Comment by siddhu001** [comment]
> > Currently, for text style: (follwing this paper https://arxiv.org/pdf/1809.09190.pdf)
> > `wav_name INTENT ENTITY SLOT_NAME slot1_value ENTITY SLOT_NAME slot2_value`
> > `01356b70-791d-42c4-b14b-56fefa67358f-1622.wav SETLIGHTCOLOR HOUSE_ROOM_UNIQUE ROOM bedroom COLOR COLOR red`
> > The next step would be training. Any suggestions about it?
> > The training set is very small, do we need a pretrained ASR?
> > Also, should we only focus on the intent classification tasks for now? (Many slot values occur infrequently in the dataset. )
> > Another dumb question, I am not sure I find the correct snips dataset, this recipe is for smart-light-en-closed-field subset, which has only 1660 utterances.
> 
> I'm not quite sure about the datasize but could you please share the link of where you get the data?

I can confirm the datasize is 1660 utterances. You have to ask for permission from the authors to access the data

File: egs2/snips/asr1/local/data.sh (hunk 5 above)
Comment (yuekaizhang) [reply egs2/snips/asr1/local/data.sh:65]: https://github.com/alexa/alexa-end-to-end-slu/blob/0d4a44edd468d9951ac9f4d0917585d3113171b2/dataprep/prepare_snips.py#L23 I would say the dataset I downloaded from snips' email has the different structure with the Amazon paper. I didn't have any .csv file. I think it is hard to know the order of utterances in the Amazon paper.

Comment (siddhu001) [reply egs2/snips/asr1/local/data.sh:65]: Hi I am sorry I was not aware the complete.csv is no longer part of Snips dataset. The collaborators with whom I worked earlier used metadata.json to create complete.csv. I think we should definitely make sure that train:valid: test split is 0.8:0.1:0.1 so as to ensure train:valid:test split size is similar to the ones used in prior work.

File: egs2/snips/asr1/run.sh (hunk 4 above)
Comment (yuekaizhang) [reply egs2/snips/asr1/run.sh:25]: For intent classification, e.g. intent INCREASEBRIGHTNESS , when you say word level decoding, you mean increase brightness (two words)? For sentence level decoding, it means treat the whole INCREASEBRIGHTNESS as one single id. Am I correct? So, what kind of decoding I should try first in your opinion?

Comment (siddhu001) [reply egs2/snips/asr1/run.sh:25]: I would actually suggest trying the whole INCREASEBRIGHTNESS as one single id since it more closely resembles a classification task. We can also try with word level decoding later.

**Comment by yuekaizhang** [comment]
Got the first result for snips intent classification task. Use enc-dec only, training from scratch, since there are only 1660 utterances, the result is ~25% acc now. Would try something  like sentence bert https://github.com/espnet/espnet/pull/3410 later.

**Comment by sw005320** [comment]
Thanks, @yuekaizhang!
- Can you paste the learning curve? The ASR number looks bad actually, but not sure this is due to the small amounts of training data or not.
- The PR looks good. So, anyway, we can merge this PR after you fix some CI errors, and continue performance tuning.

**Comment by yuekaizhang** [comment]
> Thanks, @yuekaizhang!
> 
> * Can you paste the learning curve? The ASR number looks bad actually, but not sure this is due to the small amounts of training data or not.
> * The PR looks good. So, anyway, we can merge this PR after you fix some CI errors, and continue performance tuning.

optim0_lr0
loss

I actually almost didn't tune the training. Since every epoch would only need about 60 steps (single GPU and just use 5 GB gpu memory), the learning rate and warm-up steps might need to be adjusted. Perhaps need to fine-tune with pretrained ASR model or post encoder.

**Comment by sw005320** [comment]
@yuekaizhang, I can merge your PR after you fix the conflict.
Then, we can continue tuning the recipe.

**Comment by yuekaizhang** [comment]
> @yuekaizhang, I can merge your PR after you fix the conflict.
> Then, we can continue tuning the recipe.

Thanks, sorry about the late reply.

**Comment by sw005320** [comment]
Thanks!
I just merged the PR.
Please continue tuning and report the progrss.
